bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/events/EventHandler.java

```diff
@@ -0,0 +1,148 @@
+     *
+     * @param typedEventSubscribers the event subscribers indexed by the event type
+     * @param typedEventFactories the event factories indexed by the event type
+     * @param safeCaller the safe caller to use
+     */
+    public EventHandler(final ConcurrentHashMap<String, CopyOnWriteArraySet<EventSubscriber>> typedEventSubscribers,
+            final Map<String, EventFactory> typedEventFactories, final SafeCaller safeCaller) {
+        this.typedEventSubscribers = typedEventSubscribers;
+        this.typedEventFactories = typedEventFactories;
+        this.safeCaller = safeCaller;
+    }
+
+    public void handleEvent(org.osgi.service.event.Event osgiEvent) {
+        logger.trace("Handle OSGi event (event: {})", osgiEvent);
+
+        Object typeObj = osgiEvent.getProperty("type");
+        Object payloadObj = osgiEvent.getProperty("payload");
+        Object topicObj = osgiEvent.getProperty("topic");
+        Object sourceObj = osgiEvent.getProperty("source");
+
+        if (typeObj instanceof String && payloadObj instanceof String && topicObj instanceof String) {
+            String typeStr = (String) typeObj;
+            String payloadStr = (String) payloadObj;
+            String topicStr = (String) topicObj;
+            String sourceStr = (sourceObj instanceof String) ? (String) sourceObj : null;
+            if (!typeStr.isEmpty() && !payloadStr.isEmpty() && !topicStr.isEmpty()) {
+                handleEvent(typeStr, payloadStr, topicStr, sourceStr);
+            }
+        } else {
+            logger.error(
+                    "The handled OSGi event is invalid. Expect properties as string named 'type', 'payload' and 'topic'. "
+                            + "Received event properties are: {}",
+                    Arrays.toString(osgiEvent.getPropertyNames()));
+        }
+    }
+
+    private void handleEvent(final String type, final String payload, final String topic,
+            final @Nullable String source) {
+        final EventFactory eventFactory = typedEventFactories.get(type);
+        if (eventFactory == null) {
+            logger.debug("Could not find an Event Factory for the event type '{}'.", type);
+            return;
+        }
+
+        final Set<EventSubscriber> eventSubscribers = getEventSubscribers(type);
+        if (eventSubscribers.isEmpty()) {
+            return;
+        }
+
+        final Event eshEvent = createESHEvent(eventFactory, type, payload, topic, source);
+        if (eshEvent == null) {
+            return;
+        }
+
+        dispatchESHEvent(eventSubscribers, eshEvent);
+    }
+
+    private Set<EventSubscriber> getEventSubscribers(String eventType) {
+        Set<EventSubscriber> eventTypeSubscribers = typedEventSubscribers.get(eventType);
+        Set<EventSubscriber> allEventTypeSubscribers = typedEventSubscribers.get(EventSubscriber.ALL_EVENT_TYPES);
+
+        Set<EventSubscriber> subscribers = new HashSet<>();
+        if (eventTypeSubscribers != null) {
```

Comment:
To get rid of the `null` warning here and in line 113, you can define `typedEventSubscribers` like this:

`private final ConcurrentHashMap<String, @Nullable CopyOnWriteArraySet<EventSubscriber>> typedEventSubscribers;`

Reply:
But it is not correct, the map should not contain a null value. We should not add wrong annotations just to workaround the missing EEAs.
You set the severity from errors to warnings some time ago, just because of this...